core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.metrics;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.client.file.FileSystemContext;
+import alluxio.heartbeat.HeartbeatExecutor;
+import alluxio.metrics.Metric;
+import alluxio.metrics.MetricsSystem;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Task that carries the client metrics information to master through heartheat. This class manages
+ * its own {@link MetricsMasterClient}.
+ *
+ * If the task fails to heartbeat to the master, it will destroy its old master client and recreate
+ * it before retrying.
+ */
+@NotThreadSafe
```

Comment:
Don't we access this from multiple threads despite it being not thread safe?

Reply:
This is a heartbeat executor, and it's only invoked once per JVM?

Reply:
Don't we also access it in the shutdown hook?

Reply:
That's true.